source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -843,85 +843,123 @@ void TransformationFlattenConditionalBranch::
-          operands.emplace_back(phi_inst->GetInOperand(2));
-        } else {
-          // The "false" branch is handled first in the OpPhi's operands; we
-          // thus provide operands to OpSelect in reverse of the order that
-          // they appear in the OpPhi.
-          operands.emplace_back(phi_inst->GetInOperand(2));
-          operands.emplace_back(phi_inst->GetInOperand(0));
-        }
-        phi_inst->SetOpcode(SpvOpSelect);
-        phi_inst->SetInOperands(std::move(operands));
-      });
+          break;
+        case 3:
+          // Similar to the 2D case.
+          if (message_.fresh_id_for_bvec3_selector() != 0) {
+            selector_operand = {SPV_OPERAND_TYPE_ID,
+                                {message_.fresh_id_for_bvec3_selector()}};
+            require_3d_boolean_vector = true;
+          }
+          break;
+        case 4:
+          // Similar to the 2D case.
+          if (message_.fresh_id_for_bvec4_selector() != 0) {
+            selector_operand = {SPV_OPERAND_TYPE_ID,
+                                {message_.fresh_id_for_bvec4_selector()}};
+            require_4d_boolean_vector = true;
+          }
+          break;
+        default:
+          assert(dimension == 4 && "Invalid vector dimension.");
+          break;
+      }
+    }
+    std::vector<opt::Operand> operands;
+    operands.emplace_back(selector_operand);
+
+    uint32_t branch_instruction_true_block_id =
+        branch_instruction.GetSingleWordInOperand(1);
+    uint32_t branch_instruction_false_block_id =
+        branch_instruction.GetSingleWordInOperand(2);
+
+    // The OpPhi takes values from two distinct predecessors.  One
+    // predecessor is associated with the "true" path of the conditional
+    // we are flattening, the other with the "false" path, but these
+    // predecessors can appear in either order as operands to the OpPhi
+    // instruction.  We determine in which order the OpPhi inputs should
+    // appear as OpSelect arguments by checking dominance of the true and
+    // false immediate successors of the header block.
+    if (branch_instruction_true_block_id == convergence_block_id) {
+      // The branch instruction's true block is the convergence block.  This
+      // means that the OpPhi's value associated with the branch instruction's
+      // block should the the "true" result of the OpSelect.
+      assert(branch_instruction_false_block_id != convergence_block_id &&
+             "Control should not reach here if both branches target the "
+             "convergence block.");
+      if (phi_inst->GetSingleWordInOperand(1) == message_.header_block_id()) {
+        operands.emplace_back(phi_inst->GetInOperand(0));
+        operands.emplace_back(phi_inst->GetInOperand(2));
+      } else {
+        assert(phi_inst->GetSingleWordInOperand(3) ==
+                   message_.header_block_id() &&
+               "The convergence block has the header block as a predecessor.");
```

Comment:
Is this message correct? I don't 100% understand it.

Reply:
I tried to phrase it better.